modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
 /*
- * PROJECT:         ReactOS kernel-mode tests
- * LICENSE:         LGPLv2+ - See COPYING.LIB in the top level directory
- * PURPOSE:         Kernel-Mode Test Suite Device Interface functions test
- * PROGRAMMER:      Filip Navara <xnavara@volny.cz>
+ * PROJECT:     ReactOS kernel-mode tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Test for Device Interface functions
+ * COPYRIGHT:   Copyright 2011 Filip Navara <xnavara@volny.cz>
+ *              Copyright 2021 Mark Jansen <mark.jansen@reactos.org>
+ *              Copyright 2021 Oleg Dubinskiy <oleg.dubinskij2013@yandex.ua>
  */
 
-/* TODO: what's with the prototypes at the top, what's with the if-ed out part? Doesn't process most results */
+/* TODO: Add IoRegisterDeviceInterface testcase */
 
 #include <kmt_test.h>
 #include <poclass.h>
 
 #define NDEBUG
 #include <debug.h>
 
-#if 0
-NTSTATUS
-(NTAPI *IoGetDeviceInterfaces_Func)(
-   IN CONST GUID *InterfaceClassGuid,
-   IN PDEVICE_OBJECT PhysicalDeviceObject OPTIONAL,
-   IN ULONG Flags,
-   OUT PWSTR *SymbolicLinkList);
-
-NTSTATUS NTAPI
-ReactOS_IoGetDeviceInterfaces(
-   IN CONST GUID *InterfaceClassGuid,
-   IN PDEVICE_OBJECT PhysicalDeviceObject OPTIONAL,
-   IN ULONG Flags,
-   OUT PWSTR *SymbolicLinkList);
-#endif /* 0 */
-
-static VOID DeviceInterfaceTest_Func()
+/* Predefined GUIDs are required for IoGetDeviceInterfaceAlias and IoOpenDeviceInterfaceRegistryKey.
+ * Only they can provide the aliases and the needed subkeys, unlike manually declared test GUIDs.
+ * Since IoRegisterDeviceInterface testcase is missing, it is not possible to register the new device interface
+ * and get an alias/key handle of it using this test. */
+/* From https://docs.microsoft.com/en-us/windows-hardware/drivers/stream/ksproperty-topology-categories */
+#define STATIC_KSCATEGORY_BRIDGE \
```

Comment:
These seem to be defined in `<ks.h>`. Is it possible to include that here instead of duplicating the definitions?

Reply:
No, it isn't. `ks.h` is not available in kmtests environment, that's the problem. I tried to include it, but then this failed to compile for me. Maybe it is possible to adapt the kmtests for using those headers somehow, but I don't know how. :frowning_face:

Reply:
What kind of errors do you get?